parser/ast.go

```diff
@@ -122,6 +122,35 @@ type Expression interface {
 	Eval() Expression
 }
 
+func ExpressionsHaveSameMeaning(a Expression, b Expression) (equal bool, err error) {
```

Comment:
What's the meaning of "Meaning"?  This function at least needs a godoc comment, but probably also a different name.  ExpressionsAreSame seems clearer to me, or ExpressionsAreIdentical.